packagedb/api.py

```diff
@@ -710,6 +710,101 @@ class PackageSetViewSet(viewsets.ReadOnlyModelViewSet):
     serializer_class = PackageSetAPISerializer
 
 
+class PurlValidateViewSet(viewsets.ViewSet):
+    """
+    Take a `purl` and check whether it's valid PackageURL or not.  
+    Optionally set `check_existence` to true to check whether the package exists in real world. 
+    
+    **Note:** As of now `check_existence` only supports `apache`, `composer`, `deb`, `gem`, 
+    `github`, `golang`, `maven`, `npm`, `nuget`and `pypi` ecosystems.
+
+    **Input example:**
+
+            {
+                "purl": "pkg:npm/foobar@12.3.1",
+                "check_existence": true,
+            }
+    
+    Response contains:
+
+    - valid
+        - True, if input PURL is a valid PackageURL.
+    - exists
+        - True, if input PURL exists in real world and `check_existence` flag is enabled.
+    """
+    def get_view_name(self):
+        return 'Validate PURL'
+
+    def list(self, request):
+        purl = request.query_params.get("purl")
+        check_existence = request.query_params.get("check_existence") or False
+
+        message_valid = "The provided PackageURL is valid."
+        message_not_valid = "The provided PackageURL is not valid."
+        message_valid_and_exists = (
+            "The provided Package URL is valid, and the package exists in the real world."
+        )
+        message_valid_but_dont_exist = (
+            "The provided PackageURL is valid but does not exist in the real world."
+        )
+        message_error_no_purl = (
+            "PackageURL (purl) is required. Please provide a PackageURL in the request."
+        )
+
+        if not purl:
+            return Response(
+                {
+                    "error": "Bad Request",
+                    "message": message_error_no_purl,
+                },
+                status=status.HTTP_400_BAD_REQUEST,
+            )
+
+        # validate purl
+        try:
+            package_url = PackageURL.from_string(purl)
+        except ValueError:
+            return Response(
+                {
+                    "valid": False,
+                    "message": message_not_valid,
+                    "purl": purl,
+                }
+            )
+
+        exists = None
```

Comment:
```suggestion
        exists = False
```

When I give an invalid purl to the endpoint (`/api/validate/?purl=pkg:npm/does-not-exist`), I get the following response:
```
{
    "valid": true,
    "exists": null,
    "message": "The provided PackageURL is valid.",
    "purl": "pkg:npm/does-not-eixst"
}
```

I expect to get the message `"The provided PackageURL is valid but does not exist in the real world."` and a value of False for `"exists"`

Reply:
To check whether the given purl exists in upstream ecosystem, we need to set the flag `check_existence` to true

`/api/validate/?purl=pkg:npm/does-not-exist&check_existence=true`
```
{
    "valid": true,
    "exists": false,
    "message": "The provided PackageURL is valid but does not exist in the real world.",
    "purl": "pkg:npm/does-not-eixst"
}
```